### DISCUSSION TITLE
Improve `authentication.requires` typings

### DISCUSSION THREAD

**Discussion by sobolevn** [encode/starlette#1662] [Potential Issue] [answered]
I've noticed that `authentication.requires` does not preserve type information about the passed function.

After applying `authentication.requires` on any function, it changes its type to `Any`.
Which is problematic: `Any` is not really safe to work with.

So, I propose to change that. From what I can tell, `@requires` does preserve the original function's signature. So, a bound typevar should be enough.

```python
from starlette.authentication import requires
from starlette.responses import PlainTextResponse
from starlette.requests import Request


@requires(['authenticated', 'admin'])
async def dashboard(request: Request) -> PlainTextResponse:
    return PlainTextResponse('Hello, world!')

reveal_type(dashboard)
# Before:
# note: Revealed type is "Any"

# After:
# note: Revealed type is "def (request: Request) -> typing.Coroutine[Any, Any, PlainTextResponse]"
```

Reference PR: https://github.com/encode/starlette/pull/1659

**Comment by Kludex** [comment, marked as answer]
The PR was accepted.
